experimental/sql/dialects/rel_ssa.py

```diff
@@ -0,0 +1,320 @@
+  """
+  Models a datatype in a relational SSA query.
+  """
+  name = "rel_ssa.datatype"
+
+
+@irdl_attr_definition
+class Int32(DataType):
+  """
+  Models a int32 type in a relational SSA query.
+
+  Example:
+
+  ```
+  !rel_ssa.int32
+  ```
+  """
+  name = "rel_ssa.int32"
+
+
+@irdl_attr_definition
+class String(DataType):
+  """
+  Models a string type in a relational SSA query, that can be either nullable or
+  not.
+
+  Example:
+
+  ```
+  !rel_ssa.string<0: !i1>
+  ```
+  """
+  name = "rel_ssa.string"
+
+  # TODO: redefine nullable as a property of all fields
+  nullable = ParameterDef(IntegerAttr)
+
+  @staticmethod
+  @builder
+  def get(val: Union[int, IntegerAttr]) -> 'String':
+    if isinstance(val, IntegerAttr):
+      return String([val])
+    return String([IntegerAttr.from_int_and_width(val, 1)])
+
+
+#===------------------------------------------------------------------------===#
+# Query types
+#===------------------------------------------------------------------------===#
+
+
+@irdl_attr_definition
+class Boolean(ParametrizedAttribute):
+  """
+  Models a type that can either be true or false to, e.g., show whether a tuple
+  is part of the result or not.
+
+  Example:
+
+  '''
+  !rel_ssa.bool
+  '''
+  """
+  name = "rel_ssa.bool"
+
+
+@irdl_attr_definition
+class SchemaElement(ParametrizedAttribute):
+  """
+  Models an element of a schema with name `elt_name` and type `elt_type`.
+
+  Example:
+  '''
+  !rel_ssa.schema_element<"id", !rel_ssa.int32>
+  '''
+  """
+  name = "rel_ssa.schema_element"
+
+  elt_name = ParameterDef(StringAttr)
+  elt_type = ParameterDef(DataType)
+
+  @staticmethod
+  @builder
+  def get(name: str, type_: DataType) -> 'SchemaElement':
+    return SchemaElement([StringAttr.from_str(name), type_])
+
+
+@irdl_attr_definition
+class Bag(ParametrizedAttribute):
+  """
+  Models a bag in a relational SSA query. The exact schema of the bag is part of
+  the type itself.
+
+  Example:
+
+  '''
+  !rel_ssa.bag<[!rel_ssa.schema_element<"id", !rel_ssa.int32>]>
+  '''
+  """
+  name = "rel_ssa.bag"
+
+  schema = ParameterDef(ArrayOfConstraint(SchemaElement))
+
+  @staticmethod
+  @builder
+  def get(types: list[DataType], names: list[str]) -> 'Bag':
+    schema_elts = [SchemaElement.get(n, t) for n, t in zip(names, types)]
+    return Bag([ArrayAttr.from_list(schema_elts)])
+
+
+#===------------------------------------------------------------------------===#
+# Expressions
+#===------------------------------------------------------------------------===#
+
+
+class Expression(Operation):
+  """
+  Interface class for all Expressions, i.e., operations that work on single
+  values.
+  """
+  ...
+
+
+@irdl_op_definition
+class Column(Expression):
+  """
+  References a specific column with name `col_name` and returns its value. The
+  name 'col_name' must be part of the schema of an encompassing operator.
+
+  Example:
+
+  '''
+  %0 : rel_ssa.int32 = rel_ssa.column() ["col_name" = "id"]
```

Comment:
ah I see that in the comment at the top:
```
Operators work on Bags whereas expressions work on
single values and are used to specify the exact behavior of operators.
```

We should be very loud about putting the essence of this comment into every op and type documentation.

I am quite unclear about the modelling at this point.
Here it seems you want to create a new column named "id" and containing int32 values, so I would expect a type like `column<"id", int32>`?
The problem I see is that `%0` is locally an `int32` in all its uses but in practice it is globally a `column<int32>` and we need to traverse SSA use-def chains to know what is what.
The MLIR printer/parser logic will really not like that.

Am I misunderstanding the modeling semantics you have with XDSL ?

I would additionally separate the verification invariants into a separate section `Verification Invariants:` in the same way you have `Example:`.

Reply:
I think this might just be a naming problem. In the `rel_impl` dialect I named this op `IndexByName` since it indexes into the tuple using a name. This will produce an `int32`, since that type of the field called `id` is `int32` in the given table.

Reply:
Other possibilities for naming this operation: `ColumnRef`, `FieldRef`, `ExtractValue`, `ExtractField`...